_overviews/hello-scala/companion-objects.md

```diff
@@ -0,0 +1,268 @@
+
+```scala
+val zenMasters = List(
+    Person("Nansen"),
+    Person("Joshu")
+)
+```
+
+This functionality comes from the use of companion objects. What happens is that when you define an `apply` method in a companion object, it has a special meaning to the Scala compiler. There’s a little syntactic sugar baked into Scala that lets you type this code:
+
+```scala
+val p = Person("Fred Flinstone")
+```
+
+and during the compilation process the compiler turns that code into this code:
+
+```scala
+val p = Person.apply("Fred Flinstone")
+```
+
+The `apply` method in the companion object acts as a [Factory Method](https://en.wikipedia.org/wiki/Factory_method_pattern), and Scala’s syntactic sugar lets you use the syntax shown, creating new class instances without using the `new` keyword.
+
+
+### Enabling that functionality
+
+To demonstrate how this feature works, here’s a class named `Person` along with an `apply` method in its companion object:
+
+```scala
+class Person {
+    var name = ""
+}
+
+object Person {
+    def apply(name: String): Person = {
+        var p = new Person
+        p.name = name
+        p
+    }
+}
+```
+
+To test this code, paste both the class and the object in the Scala REPL at the same time using this technique:
+
+- Start the Scala REPL from your command line (with the `scala` command)
+- Type `:paste` and press the [Enter] key
+- The REPL should respond with this text:
+
+````
+// Entering paste mode (ctrl-D to finish)
+````
+
+- Now paste both the class and object into the REPL at the same time
+- Press Ctrl-D to finish the “paste” process
+
+When that process works you should see this output in the REPL:
+
+````
+defined class Person
+defined object Person
+````
+
+>The REPL requires that a class and its companion object be entered at the same time with this technique.
+
+Now you can create a new instance of the `Person` class like this:
+
+```scala
+val p = Person.apply("Fred Flinstone")
+```
+
+That code directly calls `apply` in the companion object. More importantly, you can also create a new instance like this:
+
+```scala
+val p = Person("Fred Flinstone")
+```
+
+and this:
+
+```scala
+val zenMasters = List(
+    Person("Nansen"),
+    Person("Joshu")
+)
+```
+
+To be clear, what happens in this process is:
+
+- You type something like `val p = Person("Fred")`
+- The Scala compiler sees that there is no `new` keyword before `Person`
+- The compiler looks for an `apply` method in the companion object of the `Person` class that matches the type signature you entered
+- If it finds an `apply` method, it uses it; if it doesn’t, you get a compiler error
+
+
+### Creating multiple constructors
+
+You can create multiple `apply` methods in a companion object to provide multiple constructors. This code shows how to create both one- and two-argument constructors:
+
+```scala
+class Person {
+    var name = ""
+    var age = 0
+}
```

Comment:
Should we use `Option`s now that we have introduced them, instead of using dummy default values?

Reply:
I made this change. I’ll leave this conversation open in case you want something different here.

Reply:
I hope you don’t mind, but I’m going to “resolve” this conversation so I can more easily see what I’ve finished and what’s left to do.